## Merge-queue failure is a cross-PR batch conflict, not a defect in this PR

The `Test (ubuntu-latest)` failure that removed this PR from the queue is a compile error in the **batched** test PR, caused by an interaction with another in-flight PR — not a problem with this change against `main`.

The batch build failed with:

```
error[E0432]: no `ArrowTypeInfo` in `dora_message::metadata`
error[E0061]: Metadata::new takes 1 argument but 2 were supplied
   note: pub fn new(timestamp: uhlc::Timestamp) -> Self
```

That signature (`Metadata::new` with a single argument, and no `ArrowTypeInfo` in `dora_message::metadata`) does **not** exist on `main` — current `main` still has `Metadata::new(timestamp, type_info)` and `ArrowTypeInfo` in `metadata`. It only appears because the merge queue co-batched this PR with an in-flight refactor that drops the `type_info` parameter from `Metadata`. That refactor updated all existing call sites but naturally couldn't update the **new** unit test added here, so the combined tree fails to compile.

This PR compiles and tests cleanly against current `main` (`cargo test -p dora-runtime`, `clippy`, `fmt` all green), and GitHub reports it as mergeable (no conflict). Two clean resolutions:

1. Land this PR **before** the `Metadata` refactor — no changes needed.
2. If the refactor lands first, the new `drop_oldest_keeps_queue_bounded` test needs a one-line update to the new metadata constructor. I'm happy to push that update once the refactor is on `main` — just let me know (or it'll come through on the next re-queue that surfaces the new signature).

The production change itself (`self.queue.retain(Option::is_some)`) is independent of the `Metadata` API and is unaffected either way.

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/binaries/runtime/src/operator/channel.rs
+++ b/binaries/runtime/src/operator/channel.rs
@@ -138,7 +138,74 @@ impl InputBuffer {
         }
 
         if dropped > 0 {
+            self.compact();
             tracing::warn!("dropped {dropped} operator inputs because event queue was too full");
         }
+    }
+
+    /// Physically remove the `None` tombstones left by [`Self::drop_oldest_inputs`].
+    ///
+    /// Without this, a stalled operator (slow `on_event`) makes the deque grow
+    /// by one `Option::None` slot per received input forever — the number of
+    /// live `Some` events stays capped, but the tombstones are only ever cleared
+    /// by `pop_front` in `send_next_queued`, which never runs while the outgoing
+    /// send is pending. That defeats the bounded-memory guarantee of the
+    /// drop-oldest policy and leaks until OOM. `retain` preserves the FIFO order
+    /// of the surviving events.
+    fn compact(&mut self) {
+        self.queue.retain(Option::is_some);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn closed_event(id: &str) -> Event {
+        Event::InputClosed {
+            id: DataId::from(id.to_string()),
+        }
+    }
+
+    // The compaction that `drop_oldest_inputs` runs after evicting over-cap
+    // inputs must *physically* remove the `None` tombstones, not just leave them
+    // in place. Otherwise a stalled consumer accumulates one tombstone per
+    // dropped input without bound (they are only ever cleared by `pop_front` in
+    // `send_next_queued`, which does not run while the outgoing send is pending)
+    // — an unbounded memory leak that defeats the drop-oldest policy. This
+    // exercises the compaction primitive directly with the metadata-free
+    // `Event::InputClosed` stand-in (`compact` only distinguishes `Some`/`None`).
+    #[test]
+    fn compact_removes_tombstones_preserving_order() {
+        let mut buffer = InputBuffer::new(BTreeMap::new());
+        buffer.queue = VecDeque::from([
+            None,
+            Some(closed_event("a")),
+            None,
+            None,
+            Some(closed_event("b")),
+            None,
+        ]);
+
+        buffer.compact();
+
+        assert_eq!(
+            buffer.queue.len(),
+            2,
+            "every `None` tombstone must be removed, not merely nulled out"
+        );
+        let ids: Vec<String> = buffer
+            .queue
+            .iter()
+            .map(|e| match e {
+                Some(Event::InputClosed { id }) => id.to_string(),
+                _ => panic!("unexpected queue entry after compaction"),
+            })
+            .collect();
+        assert_eq!(
+            ids,
+            ["a", "b"],
+            "surviving events must keep their FIFO order"
+        );
     }
 }

